Variadic type variables, take 1

hishamhm opened the pull request:
This PR contains a relevant bugfix, and a new feature:

* **bugfix**: the resolution of type variables in generics now avoids clashes when resolving. When using values that contain type arguments (e.g. `function<A>(A): A`) the compiler now internally performs an alpha-conversion on each use (i.e. converting it to `function<A@123>(A@123): A@123`). This fixes odd bugs that users would sometimes see when passing generic functions as arguments to other generic functions (names would clash), and in the worst case accidentally building circular types that would hang the compiler. For example, this fixes bug #442.
* **feature**: the resolution of argument and return tuples was refactored to be consistent between function calls and function variable comparisons, and while doing it, support for _variadic type variables_ was added. A variadic type variable is a second-class type variable: it can only be used as the type of `...`, and therefore only used in function declarations. Adding ellipsis in the type variable _declaration_ marks it as variadic; the ellipsis is not used in the use of the variable (see examples below).

With variadic type variables, `pcall` and `xpcall` are no longer implemented with compiler magic (the function `special_pcall_xpcall` is removed in this PR). The type of these functions can now be represented in plain Teal:

```
global pcall: function<A..., B...>(function(A):(B), A): boolean, B
global xpcall: function<A..., B...>(function(A):(B), function(any), A): boolean, B
```

The code changes introduced by this PR are extensive, so testing from the community are especially welcome!!

hishamhm commented:
@lewis6991 this PR implements the feature you were asking for! If you could give it a spin, it would be most welcome!

euclidianAce commented:
I have had a bunch of use cases for this, will definitely try it out and report back!

euclidianAce commented:
After a very brief messing around with this branch, here are my initial thoughts:

 - This only seems to cover the "pattern matching" case for generics in function declarations, for example:
```lua
local function wrapper<Args..., Rets...>(fn: (function(Args): Rets), ...: Args): function(): Rets
   local results = { fn(...) }
   return function(): Rets
      return table.unpack(results)
   end
end

local function f(s: string): string, number, integer
   return s, 1.2, 2
end

local x = wrapper(f, "foo")
print(x())
```
Works great, but you can't seem to declare a specific type of variadic generic:
```lua
local type MyFnType = function<T...>(): T

-- This is fine
local fn: MyFnType<string> = function(): number
   return 10
end

-- This is not
local fn2: MyFnType<number, string> = function(): number, string end
```
The actual error is that `MyFnType` takes only 1 type var. Additionally, multiple variadic vars further exacerbates this issue:
```lua
local type MyFnType = function<Args..., Rets...>(Args): Rets
```
Which is a valid declaration, but can not actually be used + how would it be used? The only syntax that I can really see being unambiguous is allowing tuples in variadic type vars
```lua
local type SpecificFnType1 = MyFnType<(string, number), (number, integer)>
-- maybe only require a tuple for multiple types?
local type SpecificFnType2 = MyFnType<(string, number), number>
local type SpecificFnType3 = MyFnType<number, number>
```

 - And the brief thing that I mentioned on matrix: being able to declare tuples tables with variadic type vars would increase the usability a whole bunch

Otherwise, huge :+1: for this feature!

lewis6991 commented:
Huge thanks for implementing this.

It's not quite working for me for no specific reason I can figure out. Here's a reduced testcase:

```lua
local record A
  type future  = function<A...>(function(A))
  type async_fun = function <A...,R...>(A): future<R>

  await: function<A...>(future<A>): A
  -- async: function<A...,R...>(function(A): R): async_fun<A,R>
  async: function<A...,R...>(function(A): R): (function(A): future<R>)
end

local foo = A.async(function(): string,string
  return 'a','b'
end)

local _ = A.async(function(): string
  local f = foo()  -- as A.future<string,string>
  local b, c = A.await(f)    -- as string,string
  return b..c
end)

return A
```

Currently gives errors:
```
2 errors:
async.d.tl:7:61: mismatch in number of type arguments
async.d.tl:17:11: cannot use operator '..' for types b and c
```

Without even adapting the code to use the variadic type variables I'm getting a lot of errors, many of which I think are incorrect so it seems the handling of polymorphic functions has changed/broken.

Here's the repo I'm running on (https://github.com/lewis6991/gitsigns.nvim), which runs clean on the latest release of tl. Simpley run `make tl-build`.

lenscas commented:
> Huge thanks for implementing this.
> 
> It's not quite working for me for no specific reason I can figure out. Here's a reduced testcase:
> 
> ```lua
> local record A
>   type future  = function<A...>(function(A))
>   type async_fun = function <A...,R...>(A): future<R>
> 
>   await: function<A...>(future<A>): A
>   -- async: function<A...,R...>(function(A): R): async_fun<A,R>
>   async: function<A...,R...>(function(A): R): (function(A): future<R>)
> end
> 
> local foo = A.async(function(): string,string
>   return 'a','b'
> end)
> 
> local _ = A.async(function(): string
>   local f = foo()  -- as A.future<string,string>
>   local b, c = A.await(f)    -- as string,string
>   return b..c
> end)
> 
> return A
> ```
> 
> Currently gives errors:
> 
> ```
> 2 errors:
> async.d.tl:7:61: mismatch in number of type arguments
> async.d.tl:17:11: cannot use operator '..' for types b and c
> ```
> 
> Without even adapting the code to use the variadic type variables I'm getting a lot of errors, many of which I think are incorrect so it seems the handling of polymorphic functions has changed/broken.
> 
> Here's the repo I'm running on (https://github.com/lewis6991/gitsigns.nvim), which runs clean on the latest release of tl. Simpley run `make tl-build`.

I'm guessing it is because you return a `future<R>` but `future` is not generic? And that causes stuff to break?

lewis6991 commented:
```lua
 type future  = function<A...>(function(A))
```
@lenscas: How is it not generic?

hishamhm commented:
Thanks for all the feedback so far! I'll go through it carefully in the coming days! Quick reply though:

> async.d.tl:7:61: mismatch in number of type arguments

This is because I haven't implemented (yet) "vararg adjustment" for type variable matching. I've been mostly testing it with (x)pcall using inference, so it did not come up.

euclidianAce commented:
I think I've found somewhere that the compiler gets in an infinite loop:
when assigning to a record field
```lua
local record Foo
   call: function<Rets...>(Foo, function(): Rets): Rets
end
Foo.call = function<Rets...>(f: Foo, func: function(): Rets): Rets
end
```
`tl check foo.tl` just hangs. I put this debugging snippet in `tl.tl` to log each function call and return + limit the execution:
```lua
if os.getenv("TL_DEBUG") then
   local max <const> = assert(tonumber(os.getenv("TL_DEBUG")), "TL_DEBUG was defined, but not a number")
   local count = 0
   debug.sethook(function(event: debug.HookEvent)
      if event == "call" or event == "tail call" or event == "return" then
         local info <const> = debug.getinfo(2)
         io.stderr:write(info.name or "<anon>", info.currentline > 0 and "@" .. info.currentline or "", " :: ", event, "\n")
         io.stderr:flush()
      else
         count = count + 100
         if count > max then
            error("Too many instructions")
         end
      end
   end, "cr", 100)
end
```
Then running `TL_DEBUG=100000 tl check foo.tl` seems to report the issue being around `find_var`:
```
TL_DEBUG=100000 tl check foo.tl
# ...
find_var_type@5359 :: return
<anon>@6199 :: tail call
find_var_type@5347 :: call
find_var@5300 :: call
find_var@5309 :: return
find_var_type@5359 :: return
<anon>@6199 :: tail call
find_var_type@5347 :: call
find_var@5300 :: call
find_var@5309 :: return
find_var_type@5359 :: return
<anon>@6199 :: tail call
find_var_type@5347 :: call
find_var@5300 :: call
lua: /home/corey/dev/tl//tl.lua:12: Too many instructions
stack traceback:
        [C]: in function 'error'
        /home/corey/dev/tl//tl.lua:12: in hook '?'
        /home/corey/dev/tl//tl.lua:5301: in upvalue 'find_var'
        /home/corey/dev/tl//tl.lua:5347: in upvalue 'find_var_type'
        /home/corey/dev/tl//tl.lua:6207: in function </home/corey/dev/tl//tl.lua:6198>
        (...tail calls...)
        /home/corey/dev/tl//tl.lua:6249: in upvalue 'tuple_comparison'
        /home/corey/dev/tl//tl.lua:6717: in upvalue 'is_a'
        /home/corey/dev/tl//tl.lua:6762: in upvalue 'assert_is_a'
        /home/corey/dev/tl//tl.lua:8155: in local 'cbkind_after'
        /home/corey/dev/tl//tl.lua:3316: in upvalue 'recurse'
        /home/corey/dev/tl//tl.lua:3130: in local 'fn'
        /home/corey/dev/tl//tl.lua:3307: in function </home/corey/dev/tl//tl.lua:3287>
        (...tail calls...)
        /home/corey/dev/tl//tl.lua:9198: in function 'tl.type_check'
        /home/corey/dev/tl//tl.lua:9580: in function 'tl.process_string'
        (...tail calls...)
        /home/corey/dev/tl/tl:1101: in field '?'
        /home/corey/dev/tl/tl:1325: in main chunk
        [C]: in ?
```

euclidianAce commented:
Oh and i believe this closes #211 as my intent for that issue was a way of interacting generically with function arguments and returns in was that this can

hishamhm commented:
I pushed some initial tweaks! Still work to do, I noticed there's an issue with type variables not shadowing other variables correctly, but with the current state, this slightly modified example (to avoid the shadowing of `A`), now works:
 
```lua
local record X
   type future  = function<A...>(function(A))
   await: function<A...>(future<A>): A
   async: function<A...,R...>(function(A): R): (function(A): future<R>)
end

local foo = X.async(function(): string,string
   return 'a','b'
end)

local _ = X.async(function(): string
   local f = foo()  -- as X.future<string,string>
   local b, c = X.await(f)    -- as string,string
   return b..c
end)

return X
```

so that's some progress! :)

@euclidianace's example still hangs in a loop, I'll look at that one next!

hishamhm commented:
@euclidianAce the freeze should be fixed! thanks for the reduced test case (I added to the suite) and for the debug hook snippet (I added it to the compiler, I've been doing such things in ad hoc ways for a long time!)

hishamhm commented:
> Here's the repo I'm running on (https://github.com/lewis6991/gitsigns.nvim), which runs clean on the latest release of tl. Simpley run make tl-build.

I have quickly tested this with the current code of the PR and it still produces errors. I haven't been able to stop and try to produce reduced error cases yet — any help on that, producing short examples that work on `master` but fail on this PR, is especially welcome!

lewis6991 commented:
Sure will do! Happy to help!

hishamhm commented:
ah, and the original example with `local record A` does work too, I fixed the shadowing issue!

lewis6991 commented:
Passes on master fails on PR:

```lua
local record a
  type future1 = function<A1>(function(A1))
  type future0 = function    (function(  ))
  type async_fun1_1 = function <A1,R1>(A1): future1<R1>
  type async_fun0_0 = function        ()  : future0

  await: function<A1>(future1<A1>): A1
  async: function       (function(  ): ()): async_fun0_0
  async: function<A1,R1>(function(A1): R1): async_fun1_1 <A1,R1>
end

local get_repo_info = a.async(function(e: string): string
  return e
end)

local _ = a.async(function()
  local _: string = a.await(get_repo_info('a'))
end)
```

```
teal/gitsigns/git.tl:17:42: wrong number of arguments (given 1, expects 0)
```

lewis6991 commented:
Here's a self contained file which uses variadic type arguments but has errors. Not fully reduced but should be helpful.

```lua
local record a
  type future = function<A...>(function(A))
  type async_fun = function <A...,R...>(A): future<R>

  await: function<A...>(future<A>): A
  async: function<A...,R...>(function(A): R): async_fun<A,R>
  wrap: function<A1,A2,R...>(function(A1,A2, function(R)), integer): async_fun <A1,A2,R>
end

local await = a.await
local async = a.async

local record Hunk end
local record GJobSpec end

local record M
  record BlameInfo end

  record Version end
  version: Version

  enable_yadm: boolean

  set_version: function(string): a.future
  run_diff   : function(string, {string}, string): a.future<{Hunk}>

  record Obj
    toplevel      : string
    gitdir        : string
    file          : string
    abbrev_head   : string

    command              : function(Obj, {string}, GJobSpec): a.future<{string}>
    update_head          : function(Obj): a.future
    unstage_file         : function(Obj): a.future
    get_show             : function(Obj, string, string): a.future
    get_show_text        : function(Obj, string): a.future<{string}>
    run_blame            : function(Obj, {string}, number): a.future<BlameInfo>
    set_file_info        : function(Obj): a.future
    ensure_file_in_index : function(Obj): a.future
    stage_hunks          : function(Obj, {Hunk}, boolean): a.future
    new                  : function(string): a.future<Obj>
  end

end

local Obj = M.Obj

local command = a.wrap(function(args: {string}, _: GJobSpec, callback: function({string}))
  callback(args)
end, 3)

local get_repo_info = async(function(e: string): string
  return e
end)

M.run_diff = async(function(_: string, _: {string}, _: string): {Hunk}
  await(command({'a', 'b'}))
  return {}
end)

M.set_version = async(function(version: string)
  local _: {string} = await(command{version})
end)

Obj.command = async(function(_self: Obj, args: {string}, spec: GJobSpec): {string}
  return await(command(args, spec))
end)

Obj.update_head = async(function(_self: Obj)
  local _: string = await(get_repo_info(''))
end)

Obj.set_file_info = async(function(self: Obj)
  local _: {string} = await(self:command({'a'}))
end)

Obj.unstage_file = async(function(self: Obj)
  await(self:command{'reset', self.file })
end)

Obj.get_show_text = async(function(self: Obj, object: string): {string}
  return await(self:command({object}))
end)

Obj.get_show = async(function(self: Obj, object: string, _: string)
  await(self:command({object}))
end)

Obj.run_blame = async(function(_self: Obj, _: {string}, _: number): M.BlameInfo
  return {}
end)

Obj.ensure_file_in_index = async(function(self: Obj)
  await(self:command{'a'})
  await(self:set_file_info())
end)

Obj.stage_hunks = async(function(self: Obj, _: {Hunk}, _: boolean)
  await(self:ensure_file_in_index())
  await(self:command({'a'}))
end)

Obj.new = a.async(function(file: string): Obj
  local self = setmetatable({} as Obj, {__index = Obj})

  self.file = file
  local _: string = await(get_repo_info('a'))
  self.abbrev_head = await(get_repo_info('a'))

  await(self:set_file_info())

  return self
end)

return M
```

euclidianAce commented:
I have one case that silently produces invalid types, but doesn't error (coincedentally it's also in an async implementation)

```lua
local record Frame<T...> end
local async: function<Ret..., Args...>(fn: (function(...: Args): Ret), ...: Args): Frame<Ret>
local await: function<Ret...>(frame: Frame<Ret>): Ret

local x, y = await(async(function(): string, number end))
print_type(x)
print_type(y)
```
```
$ tl check test.tl
<invalid type>
<invalid type>
========================================
2 warnings:
test.tl:6:12: type is: <invalid type>
test.tl:7:12: type is: <invalid type>
========================================
2 errors:
test.tl:6:1: unknown variable: print_type
test.tl:7:1: unknown variable: print_type
```

Without the `print_type` it compiles fine, and even worse if you mess up the annotation it works fine:
```
local x, y: thread, table = await(async(function(): string, number end))
```

```
$ tl check test.tl
thread
{<any type> : <any type>}
========================================
2 warnings:
test.tl:6:12: type is: thread
test.tl:7:12: type is: {<any type> : <any type>}
========================================
2 errors:
test.tl:6:1: unknown variable: print_type
test.tl:7:1: unknown variable: print_type
```

kikito commented:
Just an edge case that might be interesting to include here:

AFAIK, in plain lua `...` can be used outside of function declarations.

It's "empty" on the entry file but on a *required* file it contains the string used to `require` it.

```
-- foo/bar.lua
return ...
```

```
-- main.lua
local foobar = require "foo.bar"
print(foobar)
```
Will print `foo.bar`

hishamhm commented:
@kikito as far as Teal's static analysis does, `...` at the top level is a vararg of strings, so that should work for Lua's dynamic behavior you described. :+1: 

@euclidianAce I just pushed a fix that should address the issues you described, thank you for the error cases!

@lewis6991 I haven't gotten around to fixing the two error cases you described above, but I have an idea of where the problem is for the latter (sure enough, in a place where I already had a `-- FIXME` comment in my PR!). For the former, the behavior regression, I couldn't spot the problem at first glance, but it has to do with the order of evaluation of polymorphic functions (if you flip the definitions of `async` around, it works) — that order seems to have changed but I don't know why yet. The ordering resolution algorithm tries to match the most exact arity first, but in that case both functions have the same number of arguments; the _arguments_ have different numbers of arguments. The algorithm is somewhat lax by design because of Lua's always-optional-arguments heritage...

hishamhm commented:
I am going to close this PR and merge the bugfix part of it, which I just backported to master. We can revisit the feature part of it (the actual variadic type variables) and their complexities at a later time!

lewis6991 commented:
That's quite unfortunate to hear. I was really looking forward to seeing this go in at some point. I'm finding myself needing to wrap functions more and more often (mostly to set up and restore contexts) and currently with Teal it requires me to use `as` annotations.

Are there any plans for this? The project appears to be on a bit of a hiatus.

Sorry, I know this is an annoying comment, but I just need to voice my opinion. Normally I would try and implement my own requests and raise a PR, but I'm not sure how feasible that would be here.

hishamhm commented:
@lewis6991 I still find variadic type variables (VTVs) to be a good idea and I would like to get them on the language. Closing this PR was more about this specific implementation than the concept. That's why I renamed it "take 1", in hopes a "take 2" happens in the future.

The implementation was getting unwieldy; every time I'd fix a test case I'd break another one, and this PR included other unrelated good things that needed to be merged already, so it was good to get everything that wasn't directly related to the VTVs cherry picked and merged in.

As for the project status, I consciously slowed it down over the last year so that people would have a stable language base to work with, and we could collect feedback on what needs to change or be added.

I want to make a new release soon-ish this year, and I'd love to have VTVs to be part of it, but I'm not sure I'll be able to do it myself. To be honest, right now I think a main difficulty is that once types get more and more complex, understanding the type checker is getting out of hand especially because, due to the lack of tagged-unions-or-similar, all types are of the type `Type` and all AST nodes are of the type `Node`, which feels almost like working with untyped Lua again :laughing: Fighting the edge cases on this PR had turned into a total slog. I'm honestly thinking about adding discriminated unions to the language and reworking the compiler itself to use them before I move further on making the type system any more complicated!

So, yeah, there's a desire to have this feature but no immediate short-term plans (I'm doing this on my spare time, after all, while juggling several projects!). But if you feel like taking the plunge and making a PR to add VTVs that works, I'd be happy to take it! I think the implementation in this PR has some good stuff that could serve as inspiration.
